### PR TITLE
NAT Gateways and Layers

### LINKED ISSUE #7: Create NAT Gateways
### Acceptance Criteria

 - [ ] One NAT Gateway per subnet/availability zone.
 - [ ] One EIP per NAT Gateway.

### LINKED ISSUE #16: Create TITAN Layers
### Acceptance Criteria

 - [ ] Create: DMZ layer
 - [ ] Create: Routing layer
 - [ ] Create: Service layer
 - [ ] Create: Data layer
 - [ ] Create: Admin layer
 - [ ] Create: Net layer
 - [ ] Export: per-layer subnet ids.
 - [ ] Export: per-layer route table ids.
 - [ ] Export: per-layer network ACL id.
 - [ ] Export: per-layer IPv6 CIDR blocks.
 - [ ] Export: per-layer IPv4 CIDR blocks.

### LINKED ISSUE #18: Manage VPC Default Route Table

### LINKED ISSUE #19: Manage VPC Main Network ACL

### LINKED ISSUE #20: Add VPC Endpoint for S3

### PR DESCRIPTION
**Pull request by naftulikay**
Closes #16. Closes #7. Closes #18. Closes #19. Closes #20.

### CODE REVIEW

**Comment by naftulikay** [comment]
Plan for thing with DMZ layer: https://gist.github.com/naftulikay/e04b017f922ac6b596a4c1220aba729e

**Comment by naftulikay** [comment]
Plan for addition of routing layer: https://gist.github.com/naftulikay/ebe19d10f9fa0b66eeaedbfa3ae94c60

**Comment by naftulikay** [comment]
Plan for addition of service layer: https://gist.github.com/naftulikay/b5cae2a96d6ccce1851e363768137630

**Comment by naftulikay** [comment]
Plan for addition of data layer: https://gist.github.com/naftulikay/e8b48f5e8c2d54b283b61c6bc8a4b9bb

**Comment by naftulikay** [comment]
Plan for addition of admin layer: https://gist.github.com/naftulikay/3dbea33b34c85fc2b1c967a80a29207c

**Comment by naftulikay** [comment]
Plan for addition of net layer: https://gist.github.com/naftulikay/cd90ea3a8732d92e006cb18b68fc67cf

**Comment by naftulikay** [comment]
- [ ] Export network ACLs from layers and to the network.

**Comment by naftulikay** [comment]
Plan at end: https://gist.github.com/naftulikay/45913850637831b1b9784527013646be

File: modules/aws/titan_layer/o.subnet.tf [hunk 1]
```diff
@@ -28,8 +28,9 @@ output "cidr_start" {
   EOF
 }
 
-output "ipv6_association_ids" {
-  value = ["${aws_subnet.default.*.ipv6_association_ids}"]
+output "ipv6_cidr_block_association_ids" {
+  # this is a bug in terraform documentation, docs say ipv6_association_id is the thing, it's not
```
Comment (asciifaceman) [inline modules/aws/titan_layer/o.subnet.tf]: is there an associated github issue we can contribute to to fix this?

Comment (naftulikay) [reply modules/aws/titan_layer/o.subnet.tf]: @asciifaceman [just reported here](https://github.com/terraform-providers/terraform-provider-aws/issues/2678)

**Comment by naftulikay** [comment]
Final plan before integration testing: https://gist.github.com/naftulikay/70dec4a991e58ff9a1c411beaa12e8da

**Comment by naftulikay** [comment]
Hopefully final plan: https://gist.github.com/naftulikay/80a80bbc067b41260578fced349c3d46

**Comment by asciifaceman** [comment]
HNNNNNNG

**Comment by naftulikay** [comment]
Able with the `examples/single` project to bring up and tear down in one step a TITAN network with 3 AZs.

**Comment by asciifaceman** [comment]
is it just me or did you make this 1000% simpler than originally

**Comment by naftulikay** [comment]
@asciifaceman this particular thing or generally? Terraform changed a lot so we have to make a lot less assumptions and you can actually grow a network over time so there's a clear upgrade path:

 - Startup: single hub with VPN back to your HQ with only one AZ
 - Growing Startup: expand hub to three AZs
 - We Need Separate Environments: start adding environments to the hubs
 - We Need Beastmode™ HA: expand production environment to span 6 AZs.

**Pull request merged**